File: apps/desktop/src/data/fixtures/targets.ts
// Static mock fixture data for Target and TargetDetail
// Matches design V3 mock data.

import type {
  Target,
  TargetDetail as TargetDetailType,
} from '@/bindings/types';

// ─── Design V3 flat fixture shape ───────────────────────────────────────────
//
// `uuid` maps each fixture entry to the stable UUIDv5 used by the spec-013/023
// backend.  When a row is selected in the list, its `uuid` is put into the URL
// search param and passed to `TargetDetailV2` so the detail loads from the real
// `target.get` backend command.

export interface TargetFixture {
  id: number;
  /** Stable UUIDv5 — matches the backend `targets` table row. */
  uuid: string;
  name: string;
  common: string;
  kind: string;
  sessions: number;
  hours: number;
  projects: number;
  warn?: boolean;
}

export const TARGETS_DATA: TargetFixture[] = [
  {
    id: 1,
    uuid: '550e8400-e29b-41d4-a716-446655440206',
    name: '(unresolved)',
    common: '',
    kind: 'unknown',
    sessions: 3,
    hours: 4.2,
    projects: 0,
    warn: true,
  },
  {
    id: 2,
    uuid: '550e8400-e29b-41d4-a716-446655440203',
    name: 'IC 1396',
    common: "Elephant's Trunk",
    kind: 'deep sky',
    sessions: 4,
    hours: 9.3,
    projects: 1,
  },
  {
    id: 3,
    uuid: '550e8400-e29b-41d4-a716-446655440204',
    name: 'Jupiter',
    common: '',
    kind: 'planetary',
    sessions: 6,
    hours: 2.5,
    projects: 1,
  },
  {
    id: 4,
    uuid: '550e8400-e29b-41d4-a716-446655440202',
    name: 'M31',
    common: 'Andromeda Galaxy',
    kind: 'deep sky',
    sessions: 8,
    hours: 11.8,
    projects: 1,
    warn: true,
  },
  {
    id: 5,
    uuid: '550e8400-e29b-41d4-a716-446655440205',
    name: 'M42',
    common: 'Orion Nebula',
    kind: 'deep sky',
    sessions: 5,
    hours: 3.4,
    projects: 0,
  },
  {
    id: 6,
    uuid: '550e8400-e29b-41d4-a716-446655440201',
    name: 'NGC 7000',
    common: 'North America Nebula',
    kind: 'deep sky',
    sessions: 12,
    hours: 14.2,
    projects: 2,
  },
];

// ---------------------------------------------------------------------------
// Rich list items — shown in the left pane (existing shape retained)
// ---------------------------------------------------------------------------

export const targets: Target[] = [
  {
    id: '550e8400-e29b-41d4-a716-446655440201',
    name: 'NGC 7000',
    aliases: ['North America Nebula', 'Caldwell 20'],
    catalogIds: { ngc: '7000' },
    kind: 'deep_sky',
    coordinates: { ra: 20.983, dec: 44.517 },
    sessionCount: 12,
    projectCount: 2,
    totalIntegrationHours: 14.2,
    coverage: { Ha: 6.3, OIII: 4.8, SII: 1.8, L: 0 },
    recommendedHours: { Ha: 6.0, OIII: 5.0, SII: 3.0, L: 0 },
  },
  {
    id: '550e8400-e29b-41d4-a716-446655440202',
    name: 'M31',
    aliases: ['Andromeda Galaxy'],
    catalogIds: { messier: '31', ngc: '224' },
    kind: 'deep_sky',
    coordinates: { ra: 0.712, dec: 41.27 },
    sessionCount: 8,
    projectCount: 1,
    totalIntegrationHours: 11.8,
    coverage: { L: 5.2, R: 3.1, G: 1.8, B: 1.7 },
    recommendedHours: { L: 6.0, R: 3.0, G: 3.0, B: 3.0 },
  },
  {
    id: '550e8400-e29b-41d4-a716-446655440203',
    name: 'IC 1396',
    aliases: ["Elephant's Trunk"],
    catalogIds: { ic: '1396' },
    kind: 'deep_sky',
    coordinates: { ra: 21.62, dec: 57.5 },
    sessionCount: 4,
    projectCount: 1,
    totalIntegrationHours: 9.3,
    coverage: { Ha: 5.1, OIII: 4.2 },
    recommendedHours: { Ha: 6.0, OIII: 6.0 },
  },
  {
    id: '550e8400-e29b-41d4-a716-446655440204',
    name: 'Jupiter',
    aliases: [],
    catalogIds: {},
    kind: 'planetary',
    coordinates: {},
    sessionCount: 6,
    projectCount: 1,
    totalIntegrationHours: 2.5,
    coverage: {},
    recommendedHours: {},
  },
  {
    id: '550e8400-e29b-41d4-a716-446655440205',
    name: 'M42',
    aliases: ['Orion Nebula'],
    catalogIds: { messier: '42', ngc: '1976' },
    kind: 'deep_sky',
    coordinates: { ra: 5.588, dec: -5.39 },
    sessionCount: 5,
    projectCount: 0,
    totalIntegrationHours: 3.4,
    coverage: { Ha: 3.4 },
    recommendedHours: { Ha: 4.0 },
  },
  {
    id: '550e8400-e29b-41d4-a716-446655440206',
    name: '(unresolved)',
    aliases: [],
    catalogIds: {},
    kind: 'deep_sky',
    coordinates: {},
    sessionCount: 3,
    projectCount: 0,
    totalIntegrationHours: 4.2,
    coverage: { Ha: 4.2 },
    recommendedHours: { Ha: 4.0 },
  },
];

// ---------------------------------------------------------------------------
// Detail — the full NGC 7000 detail matching the wireframe
// ---------------------------------------------------------------------------

export const targetDetail: TargetDetailType = {
  ...targets[0],
  sessions: [
    {
      id: 's-001',
      sessionKey: {
        target: 'NGC 7000',
        filter: 'Ha',
        binning: '1x1',
        gain: '100',
        night: '2024-11-30',
      },
      confidence: 'confirmed',
      opticalTrainId: 'ot-2600mm',
      frameCount: 54,
      totalIntegrationSeconds: 16200,
      totalSizeBytes: 0,
      metadata: {},
      targetIds: ['550e8400-e29b-41d4-a716-446655440201'],
      projectIds: ['proj-hoo'],
      warnings: [],
    },
    {
      id: 's-002',
      sessionKey: {
        target: 'NGC 7000',
        filter: 'OIII',
        binning: '1x1',
        gain: '100',
        night: '2024-11-30',
      },
      confidence: 'confirmed',
      opticalTrainId: 'ot-2600mm',
      frameCount: 38,
      totalIntegrationSeconds: 11520,
      totalSizeBytes: 0,
      metadata: {},
      targetIds: ['550e8400-e29b-41d4-a716-446655440201'],
      projectIds: ['proj-hoo'],
      warnings: [],
    },
    {
      id: 's-003',
      sessionKey: {
        target: 'NGC 7000',
        filter: 'SII',
        binning: '1x1',
        gain: '100',
        night: '2024-12-01',
      },
      confidence: 'high',
      opticalTrainId: 'ot-2600mm',
      frameCount: 22,
      totalIntegrationSeconds: 6480,
      totalSizeBytes: 0,
      metadata: {},
      targetIds: ['550e8400-e29b-41d4-a716-446655440201'],
      projectIds: [],
      warnings: [],
    },
    {
      id: 's-004',
      sessionKey: {
        target: 'NGC 7000',
        filter: 'Ha',
        binning: '1x1',
        gain: '100',
        night: '2024-12-15',
      },
      confidence: 'confirmed',
      opticalTrainId: 'ot-2600mm',
      frameCount: 30,
      totalIntegrationSeconds: 9000,
      totalSizeBytes: 0,
      metadata: {},
      targetIds: ['550e8400-e29b-41d4-a716-446655440201'],
      projectIds: ['proj-sho'],
      warnings: [],
    },
  ],
  projects: [
    { id: 'proj-hoo', name: 'NGC 7000 · HOO', state: 'processing' },
    { id: 'proj-sho', name: 'NGC 7000 · SHO mosaic', state: 'ready' },
  ],
};

File: apps/desktop/src-tauri/src/commands/targets.rs
//! Spec 029 target stubs exposed to the Tauri webview.
//!
//! Stub implementations returning hardcoded fixture data matching the mock
//! layer until the real persistence layer is wired.

use std::collections::HashMap;

use contracts_core::lifecycle::ProjectState;
use contracts_core::sessions::{
    AcquisitionSession, ConfidenceLevel, MetaValue, ProvenanceOrigin, SessionKey, SessionState,
};
use contracts_core::targets::{
    CatalogIds, Coordinates, Target, TargetDetail, TargetKind, TargetProjectStub,
};
use contracts_core::ContractError;
use contracts_core::JsonAny;

/// `targets.list` — returns all targets, optionally filtered by search.
///
/// # Errors
/// Returns `Err(String)` on failure; the stub never fails.
#[tauri::command]
#[specta::specta]
pub async fn targets_list(search: Option<String>) -> Result<Vec<Target>, ContractError> {
    tracing::debug!("stub: targets.list search={search:?}");
    let targets = stub_targets();
    if let Some(q) = search {
        let q = q.to_lowercase();
        Ok(targets.into_iter().filter(|t| t.name.to_lowercase().contains(&q)).collect())
    } else {
        Ok(targets)
    }
}

/// `targets.get` — returns a single target detail.
///
/// # Errors
/// Returns `Err(String)` on failure; the stub never fails.
#[tauri::command]
#[specta::specta]
pub async fn targets_get(id: String) -> Result<TargetDetail, ContractError> {
    tracing::debug!("stub: targets.get id={id}");
    let base = stub_targets()
        .into_iter()
        .next()
        .ok_or_else(|| ContractError::internal("no stub target available"))?;
    Ok(TargetDetail {
        id: id.clone(),
        name: base.name,
        aliases: base.aliases,
        catalog_ids: base.catalog_ids,
        kind: base.kind,
        coordinates: base.coordinates,
        session_count: base.session_count,
        project_count: base.project_count,
        total_integration_hours: base.total_integration_hours,
        coverage: base.coverage,
        recommended_hours: base.recommended_hours,
        sessions: vec![AcquisitionSession {
            id: "550e8400-e29b-41d4-a716-446655440001".to_owned(),
            session_key: SessionKey {
                target: "NGC 7000".to_owned(),
                filter: "Ha".to_owned(),
                binning: "1".to_owned(),
                gain: "100".to_owned(),
                night: "2026-04-12".to_owned(),
            },
            state: SessionState::Discovered,
            confidence: ConfidenceLevel::Medium,
            optical_train_id: "550e8400-e29b-41d4-a716-446655440101".to_owned(),
            frame_count: 18,
            total_integration_seconds: 10800.0,
            total_size_bytes: 1_258_291_200,
            metadata: HashMap::from([(
                "target".to_owned(),
                MetaValue {
                    value: JsonAny::from(serde_json::json!("NGC 7000")),
                    raw: Some("NGC7000".to_owned()),
                    origin: ProvenanceOrigin::Observed,
                    confidence: ConfidenceLevel::Medium,
                    evidence_ref: None,
                },
            )]),
            target_ids: vec![id.clone()],
            project_ids: vec![],
            warnings: vec![],
        }],
        projects: vec![
            TargetProjectStub {
                id: "proj-001".to_owned(),
                name: "NGC 7000 Narrowband".to_owned(),
                state: ProjectState::Processing,
            },
            TargetProjectStub {
                id: "proj-002".to_owned(),
                name: "NGC 7000 LRGB".to_owned(),
                state: ProjectState::Ready,
            },
        ],
    })
}

// ---------------------------------------------------------------------------
// Fixture data
// ---------------------------------------------------------------------------

fn stub_targets() -> Vec<Target> {
    vec![
        Target {
            id: "550e8400-e29b-41d4-a716-446655440201".to_owned(),
            name: "NGC 7000".to_owned(),
            aliases: vec!["North America Nebula".to_owned(), "Caldwell 20".to_owned()],
            catalog_ids: CatalogIds { ngc: Some("7000".to_owned()), ic: None, messier: None },
            kind: TargetKind::DeepSky,
            coordinates: Some(Coordinates { ra: Some(314.68), dec: Some(44.31) }),
            session_count: 5,
            project_count: 2,
            total_integration_hours: 12.5,
            coverage: HashMap::from([
                ("Ha".to_owned(), 6.0),
                ("OIII".to_owned(), 4.5),
                ("SII".to_owned(), 2.0),
            ]),
            recommended_hours: HashMap::from([
                ("Ha".to_owned(), 10.0),
                ("OIII".to_owned(), 8.0),
                ("SII".to_owned(), 8.0),
            ]),
        },
        Target {
            id: "550e8400-e29b-41d4-a716-446655440202".to_owned(),
            name: "M31".to_owned(),
            aliases: vec!["Andromeda Galaxy".to_owned()],
            catalog_ids: CatalogIds {
                ngc: Some("224".to_owned()),
                ic: None,
                messier: Some("31".to_owned()),
            },
            kind: TargetKind::DeepSky,
            coordinates: Some(Coordinates { ra: Some(10.68), dec: Some(41.27) }),
            session_count: 3,
            project_count: 1,
            total_integration_hours: 8.0,
            coverage: HashMap::from([
                ("L".to_owned(), 4.0),
                ("R".to_owned(), 1.5),
                ("G".to_owned(), 1.5),
                ("B".to_owned(), 1.0),
            ]),
            recommended_hours: HashMap::from([
                ("L".to_owned(), 10.0),
                ("R".to_owned(), 4.0),
                ("G".to_owned(), 4.0),
                ("B".to_owned(), 4.0),
            ]),
        },
        Target {
            id: "550e8400-e29b-41d4-a716-446655440203".to_owned(),
            name: "IC 1396".to_owned(),
            aliases: vec!["Elephant Trunk Nebula".to_owned()],
            catalog_ids: CatalogIds { ngc: None, ic: Some("1396".to_owned()), messier: None },
            kind: TargetKind::DeepSky,
            coordinates: Some(Coordinates { ra: Some(324.75), dec: Some(57.49) }),
            session_count: 1,
            project_count: 0,
            total_integration_hours: 2.0,
            coverage: HashMap::from([("SII".to_owned(), 2.0)]),
            recommended_hours: HashMap::from([
                ("Ha".to_owned(), 10.0),
                ("OIII".to_owned(), 8.0),
                ("SII".to_owned(), 8.0),
            ]),
        },
    ]
}
